Cython/Utility/ImportExport.c

```diff
@@ -1,178 +1,117 @@
+            Py_DECREF(listitem);
+            if (!hasattr) {
+                goto not_found;
+            }
         }
-        Py_DECREF(spec);
-        Py_DECREF(module);
-    } else if (PyErr_Occurred()) {
-        PyErr_Clear();
+        *module = imported_module;
+        return 1;
     }
-#endif
-
-    return __Pyx__ImportDottedModule(name, parts_tuple);
-}
-
 
-/////////////// ImportDottedModuleRelFirst.proto ///////////////
-
-static PyObject *__Pyx_ImportDottedModuleRelFirst(PyObject *name, PyObject *parts_tuple); /*proto*/
-
-/////////////// ImportDottedModuleRelFirst ///////////////
-//@requires: ImportDottedModule
-//@requires: Import
+    // Get top-level module
+    list = PyUnicode_Split(qualname, PYUNICODE("."), 1);
+    if (unlikely(!list)) {
+        goto error;
+    }
 
-static PyObject *__Pyx_ImportDottedModuleRelFirst(PyObject *name, PyObject *parts_tuple) {
-    PyObject *module;
-    PyObject *from_list = NULL;
-    module = __Pyx_Import(name, from_list, -1);
-    Py_XDECREF(from_list);
-    if (module) {
-        if (parts_tuple) {
-            module = __Pyx_ImportDottedModule_WalkParts(module, name, parts_tuple);
-        }
-        return module;
+    listsize = __Pyx_PyList_GET_SIZE(list);
+    #if !CYTHON_ASSUME_SAFE_SIZE
+    if (unlikely(listsize == -1)) {
+        goto error;
     }
-    if (unlikely(!PyErr_ExceptionMatches(PyExc_ImportError)))
-        return NULL;
-    PyErr_Clear();
-    // try absolute import
-    return __Pyx_ImportDottedModule(name, parts_tuple);
-}
+    #endif
+    if (listsize == 1) {
+        // No dot, so we already have the top-level module
+        *module = imported_module;
+        return 1;
+    }
+    Py_DECREF(imported_module);
 
+    listitem = __Pyx_PyList_GET_ITEM(list, 0);
+    #if !CYTHON_ASSUME_SAFE_MACROS
+    if (unlikely(!listitem)) {
+        *module = NULL;
+        return -1;
+    }
+    #endif
 
-/////////////// Import.proto ///////////////
+    status = __Pyx__Import_GetModule(listitem, module);
+    Py_DECREF(list);
+    return status;
 
-static PyObject *__Pyx_Import(PyObject *name, PyObject *from_list, int level); /*proto*/
+error:
+    Py_DECREF(imported_module);
+    *module = NULL;
+    return -1;
 
-/////////////// Import ///////////////
-//@requires: ObjectHandling.c::PyObjectGetAttrStr
-//@requires:StringTools.c::IncludeStringH
+not_found:
+    Py_DECREF(imported_module);
+    *module = NULL;
+    return 0;
+}
 
-static PyObject *__Pyx_Import(PyObject *name, PyObject *from_list, int level) {
+static PyObject *__Pyx_Import(PyObject *name, PyObject *from_list, PyObject *qualname, int level) {
     PyObject *module = 0;
     PyObject *empty_dict = 0;
     PyObject *empty_list = 0;
+    int status;
+
+    if (!qualname) {
+        qualname = name;
+    }
+    status = __Pyx__Import_Lookup(qualname, from_list, &module);
+    if (unlikely(status == -1)) {
+        return NULL;
+    }
+    else if (likely(status == 1)) {
+        return module;
+    }
```

Comment:
The C compiler might do it for us, but also visually, it seems more natural to me to put the "likely" case first.
```suggestion
    module_found = __Pyx__Import_Lookup(qualname, from_list, &module);
    if (likely(module_found == 1)) {
        return module;
    } else if (unlikely(module_found == -1)) {
        return NULL;
    }
```